From 38003f0f7211c382c0c72b2eb9ade198f346c9aa Mon Sep 17 00:00:00 2001
From: Vinith Kumar R <quic_vinithku@quicinc.com>
Date: Mon, 12 Aug 2024 23:20:33 +0530
Subject: [PATCH] QSDK: wifi: Disable warn_on and error dump for nwifi mode

Currently, for native wifi mode, if packet header length exceeds the
threshold value 30, then warn_on will hit. With this handling, facing
panic_on_warn_on for the corrupted packet due to header length
exceeding the threshold value. So, as a workaround, disabling warn_on
call and error dump with the module param flag for native wifi mode.

Patch-dependency: 683-ath12k-Debug-prints-to-prevent-buffer-overflow.patch

Signed-off-by: Vinith Kumar R <quic_vinithku@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c  |  5 +++++
 drivers/net/wireless/ath/ath12k/core.h  |  1 +
 drivers/net/wireless/ath/ath12k/dp_rx.c | 14 +++++++++-----
 3 files changed, 15 insertions(+), 5 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.c b/drivers/net/wireless/ath/ath12k/core.c
index 7b3b863..12fe733 100644
--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -87,6 +87,11 @@ module_param_array_named(rfs_core_mask, ath12k_rfs_core_mask, int, NULL, 0644);
 MODULE_PARM_DESC(rfs_core_mask, "Default RFS core mask, mask for 2G, mask for 5G,\n"
 		 "mask for 6G. One bit for one CPU core\n");
 
+bool ath12k_rx_nwifi_err_dump = false;
+module_param_named(rx_nwifi_err_dump, ath12k_rx_nwifi_err_dump, bool, 0644);
+MODULE_PARM_DESC(rx_nwifi_err_dump,
+                 "rx nwifi err dump: 0-disable, 1-enable");
+
 static DEFINE_MUTEX(ath12k_hw_lock);
 static struct list_head ath12k_hw_groups = LIST_HEAD_INIT(ath12k_hw_groups);
 
diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index 52d9d36..7e043c2 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -55,6 +55,7 @@
 extern unsigned int ath12k_frame_mode;
 extern bool ath12k_erp_support;
 extern unsigned int ath12k_ppe_ds_enabled;
+extern bool ath12k_rx_nwifi_err_dump;
 
 #define ATH12K_MAX_CORE_MASK	(0xFFFF & ((1 << NR_CPUS) - 1))
 extern unsigned int ath12k_rfs_core_mask[4];
diff --git a/drivers/net/wireless/ath/ath12k/dp_rx.c b/drivers/net/wireless/ath/ath12k/dp_rx.c
index 4c6a206..7e90189 100644
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -3283,10 +3283,12 @@ static bool ath12k_dp_rx_check_max_nwifi_hdr_len(struct ath12k_base *ab,
 	hdr_len = ieee80211_hdrlen(hdr->frame_control);
 	if (unlikely(hdr_len > DP_MAX_NWIFI_HDR_LEN)) {
 		ab->soc_stats.invalid_rbm++;
-		ath12k_err_dump(ab, msg, "msdu_data: ", msdu->data, msdu->len,
-				NULL);
-		ath12k_err_dump(ab, NULL, "rx_desc: ", rx_desc, sizeof(*rx_desc),
-				NULL);
+		if (ath12k_rx_nwifi_err_dump) {
+			ath12k_err_dump(ab, msg, "msdu_data: ", msdu->data, msdu->len,
+					NULL);
+			ath12k_err_dump(ab, NULL, "rx_desc: ", rx_desc, sizeof(*rx_desc),
+					NULL);
+		}
 		return true;
 	}
 
@@ -3359,7 +3361,9 @@ static int ath12k_dp_rx_process_msdu(struct ath12k *ar,
 	if (rx_info->decap_type == DP_RX_DECAP_TYPE_NATIVE_WIFI &&
 	    ath12k_dp_rx_check_max_nwifi_hdr_len(ab, rx_desc, msdu,
 						 "Invalid len in Rx\n")) {
-		WARN_ON_ONCE(1);
+		if( ath12k_rx_nwifi_err_dump )
+			WARN_ON_ONCE(1);
+
 		ret = -EINVAL;
 		goto free_out;
 	}
-- 
2.34.1

